server/src/main/java/org/elasticsearch/indices/recovery/plan/ShardSnapshotsService.java

```diff
@@ -91,7 +93,13 @@ public void fetchLatestSnapshotsForShard(ShardId shardId, ActionListener<Optiona
 
         logger.debug("Searching for peer recovery compatible snapshots in [{}]", repositories);
 
-        GetShardSnapshotRequest request = GetShardSnapshotRequest.latestSnapshotInRepositories(shardId, repositories);
+        GetShardSnapshotRequest request = GetShardSnapshotRequest.latestSnapshotInRepositories(
+            clusterService.state().getMinTransportVersion().onOrAfter(TransportVersions.SNAPSHOT_REQUEST_TIMEOUTS)
+                ? TimeValue.MINUS_ONE
+                : TimeValue.MAX_VALUE,
```

Comment:
This is an internal request so really should have infinite timeout (as long as the cluster is new enough to understand that - using the current PR's transport version to be sure). Previously it had a 30s timeout which was probably a mistake.